source/common/http/header_map_impl.cc

```diff
@@ -21,190 +21,103 @@ constexpr size_t MinDynamicCapacity{32};
+// Initialize as a Type::Reference
+HeaderString::HeaderString(absl::string_view ref_value)
+    : buffer_(ref_value), string_length_(ref_value.size()) {
   ASSERT(valid());
 }
 
-HeaderString::HeaderString(HeaderString&& move_value) noexcept {
-  type_ = move_value.type_;
-  string_length_ = move_value.string_length_;
-  switch (move_value.type_) {
-  case Type::Reference: {
-    buffer_.ref_ = move_value.buffer_.ref_;
-    break;
-  }
-  case Type::Dynamic: {
-    // When we move a dynamic header, we switch the moved header back to its default state (inline).
-    buffer_.dynamic_ = move_value.buffer_.dynamic_;
-    dynamic_capacity_ = move_value.dynamic_capacity_;
-    move_value.type_ = Type::Inline;
-    move_value.buffer_.dynamic_ = move_value.inline_buffer_;
-    move_value.clear();
-    break;
-  }
-  case Type::Inline: {
-    buffer_.dynamic_ = inline_buffer_;
-    memcpy(inline_buffer_, move_value.inline_buffer_, string_length_);
-    move_value.string_length_ = 0;
-    break;
-  }
-  }
+HeaderString::HeaderString(HeaderString&& move_value) noexcept
+    : buffer_(move_value.buffer_), string_length_(move_value.string_length_) {
+  move_value.clear();
   ASSERT(valid());
 }
 
-HeaderString::~HeaderString() { freeDynamic(); }
-
-void HeaderString::freeDynamic() {
-  if (type_ == Type::Dynamic) {
-    free(buffer_.dynamic_);
-  }
-}
+HeaderString::~HeaderString() {}
 
 bool HeaderString::valid() const { return validHeaderString(getStringView()); }
 
+#define BUFFER_STR_VIEW_GET absl::get<absl::string_view>(buffer_)
+#define BUFFER_IN_VEC_GET absl::get<absl::InlinedVector<char, 128>>(buffer_)
+
 void HeaderString::append(const char* data, uint32_t size) {
-  switch (type_) {
-  case Type::Reference: {
-    // Rather than be too clever and optimize this uncommon case, we dynamically
-    // allocate and copy.
-    type_ = Type::Dynamic;
-    const uint64_t new_capacity = newCapacity(string_length_, size);
-    if (new_capacity > MinDynamicCapacity) {
-      validateCapacity(new_capacity);
-      dynamic_capacity_ = new_capacity;
-    } else {
-      dynamic_capacity_ = MinDynamicCapacity;
-    }
-    char* buf = static_cast<char*>(malloc(dynamic_capacity_));
-    RELEASE_ASSERT(buf != nullptr, "");
-    memcpy(buf, buffer_.ref_, string_length_);
-    buffer_.dynamic_ = buf;
-    break;
+  uint64_t new_capacity = static_cast<uint64_t>(size) + string_length_;
+  if (new_capacity < MinDynamicCapacity) {
+    new_capacity = MinDynamicCapacity;
   }
+  validateCapacity(new_capacity);
+  ASSERT(validHeaderString(absl::string_view(data, size)));
 
-  case Type::Inline: {
-    const uint64_t new_capacity = static_cast<uint64_t>(size) + string_length_;
-    if (new_capacity <= sizeof(inline_buffer_)) {
-      // Already inline and the new value fits in inline storage.
-      break;
-    }
-
-    FALLTHRU;
+  if (type() == Type::Reference) {
+    // Rather than be too clever and optimize this uncommon case, we switch to
+    // Inline mode and copy.
```

Comment:
A little confused. If the reference string is larger than the inline vector capacity, why are we switching to inline mode?
Is there a test for that case? That might be interesting.

Reply:
OK! I had a misunderstanding. Turns out when absl::InlinedVector<char, N> exceeds N it auto-magically allocates more space